Mozilla 对行业最大的贡献是 Rust，而不是火狐浏览器 (#8464)
xitu/gold-miner: article/2021/rust-not-firefox-is-mozillas-greatest-industry-contribution.md

> * 原文地址：[Rust, not Firefox, is Mozilla's greatest industry contribution](https://www.techrepublic.com/article/rust-not-firefox-is-mozillas-greatest-industry-contribution/)
> * 原文作者：[Matt Asay](https://www.techrepublic.com/meet-the-team/us/matt-asay/)
> * 译文出自：[掘金翻译计划](https://github.com/xitu/gold-miner)
> * 本文永久链接：[https://github.com/xitu/gold-miner/blob/master/article/2021/rust-not-firefox-is-mozillas-greatest-industry-contribution.md](https://github.com/xitu/gold-miner/blob/master/article/2021/rust-not-firefox-is-mozillas-greatest-industry-contribution.md)
> * 译者：[灰灰 greycodee](https://github.com/greycodee)

> > * 校对者：

Author: ```suggestion
> * 校对者：[PingHGao](https://github.com/PingHGao)、[Chorer](https://github.com/Chorer)、[PassionPenguin](https://github.com/PassionPenguin)
```

> * 原文作者：Matt Asay
> * 译文出自：[掘金翻译计划](https://github.com/xitu/gold-miner)
> * 本文永久链接：[https://github.com/xitu/gold-miner/blob/master/article/2021/rust-not-firefox-is-mozillas-greatest-industry-contribution.md](https://github.com/xitu/gold-miner/blob/master/article/2021/rust-not-firefox-is-mozillas-greatest-industry-contribution.md)
> * 译者：[灰灰 greycodee](https://github.com/greycodee)
> * 校对者：

# Mozilla 对行业最大的贡献是 Rust，而不是火狐浏览器

Linus Torvalds 是著名的 Linux 创造者，同时他创造的 Git 也对世界产生了巨大的影响。同样，创造火狐浏览器的 Mozilla 也开发了 Rust 语言，这对开发行业产生了深远的影响。

## Mozilla：寻求新目标

> Mozilla 看见了更好的日子。曾经有一段时间，网络自由是必不可少的。那时微软的 Internet Explorer 是主要的 Web 浏览器，而真正的担忧是 Web 的未来，其主要网关由一家大型专有公司拥有。多年以来，Mozilla 在创造一个更加开发和自由的网络环境上取得了巨大的成功。不幸的是，成功的果实却被谷歌的 Chrom 浏览器摘取了。多年后，我们用一个霸权换另一个霸权，但是[火狐浏览器却被排挤在外了](https://twitter.com/mjasay/status/902652369607036928?lang=en)。

Reviewer: ```suggestion
Mozilla 曾辉煌一时。曾几何时，它对于网络自由是不可或缺的。那时微软的 Internet Explorer 是主要的 Web 浏览器，这不由得令人担忧 Web 的未来，因为其主要入口被一家大型私有公司所掌控。经过多年的努力，Mozilla 在创造一个更加开放和自由的网络环境上大获成功。不幸的是，成功的果实却被谷歌的 Chrome 浏览器窃取。多年后，掌控大权的浏览器发生更替，而[火狐浏览器却被排挤在外了](https://twitter.com/mjasay/status/902652369607036928?lang=en)。
```

这里的 gateway 和网关没有任何关系，作者说的是流量入口。

> 尽管 Mozilla 奋斗了十年来寻找新的目标，情况还是比较糟糕。或许[ Mozilla 可以创建下一个伟大的平台](https://www.cnet.com/news/forget-facebook-the-webs-platform-is-firefox/)（不会的）。或是一个[伟大的手机操作系统](https://www.cnet.com/news/why-the-death-of-the-firefox-phone-matters/)（也不太可能）? [同步](https://twitter.com/mjasay/status/239069091573420032)（再次说不）？很多希望和错误的开始，导致不可避免的『不』。在 2017 年[ CNET 采访了 Mozilla 当时的首席执行官 Chris Beard ](https://www.cnet.com/news/mozilla-three-years-later-is-firefox-in-a-better-place/)，从他们的前景来看，并不是很乐观。

Reviewer: ```suggestion
尽管 Mozilla 奋斗了十年来寻找新的战略目标，情况还是比较糟糕。或许 [Mozilla 可以创建下一个伟大的平台](https://www.cnet.com/news/forget-facebook-the-webs-platform-is-firefox/)（不行）。或是一个[伟大的手机操作系统](https://www.cnet.com/news/why-the-death-of-the-firefox-phone-matters/)（也不行）？ [同步](https://twitter.com/mjasay/status/239069091573420032)（同样不行）？过多的希望和错误的开始，最终必然导致“不行”。在 2017 年 [CNET 采访了 Mozilla 当时的首席执行官 Chris Beard](https://www.cnet.com/news/mozilla-three-years-later-is-firefox-in-a-better-place/)，试图解读 Mozilla 的前景，但他们自己似乎也很迷茫。
```

> 然而，在所有这些斗争中，Mozilla创造了一个真正伟大的东西：Rust。

Reviewer: ```suggestion
然而，在所有这些努力中，Mozilla 创造了一个真正伟大的东西：Rust。
```

## 平静的 Rust

从某种意义上说，十年前从 Mozilla 的研究领域中出现了一种系统编程语言是很奇怪的。之所以奇怪，是因为像什么手机浏览器、邮件客户端、手机系统等等，都是比较好的东西，而公司却去创建一个编程语言，虽然可能对创造安全的浏览器组件有点用处，但是却不一定能给 Mozilla 带来光明的未来。 

Rust 起始于 2006 年 Mozilla 员工 Graydon Hoare 的一个个人项目。[Hoare 解释](https://www.infoq.com/news/2012/08/Interview-Rust/)了他2012年工作背后的原因：

> > 许多显而易见的好想法，在其他语言中广为人知和喜爱，还没有使其成为广泛使用的系统语言，或者以内存模型非常差（不安全，对并发性敌对）的语言为由而没有进行部署。在70年代末和80年代初，在这个领域有很多优秀的竞争对手，我想根据环境已经改变的理论，复兴他们的一些想法，再给他们一个机会：Internet 是高度并发且有高度的安全意识，因此，用 C 和 C ++ 的设计的方案正在发生变化。

Reviewer 2: ```suggestion
> 许多显而易见的好想法，在其他语言中被广泛接受和喜爱，却没有加入到被广泛使用的系统语言中，或者只在内存模型非常差（不安全，对并发性敌对）的系统语言中加入。在70年代末和80年代初，在这个领域有很多优秀的竞争对手，我想根据环境已经改变的理论，复兴他们的一些想法，再给他们一个机会：Internet 是高度并发且有高度的安全意识，因此，用 C 和 C ++ 的设计的方案正在发生变化。
```

> > 许多显而易见的好想法，在其他语言中广为人知和喜爱，还没有使其成为广泛使用的系统语言，或者以内存模型非常差（不安全，对并发性敌对）的语言为由而没有进行部署。在70年代末和80年代初，在这个领域有很多优秀的竞争对手，我想根据环境已经改变的理论，复兴他们的一些想法，再给他们一个机会：Internet 是高度并发且有高度的安全意识，因此，用 C 和 C ++ 的设计的方案正在发生变化。

Reviewer 2: ```suggestion
> 许多显而易见的好想法，在其他语言中广为人知和喜爱，还没有使其成为广泛使用的系统语言，或者以内存模型非常差（不安全，对并发性敌对）的语言为由而没有进行部署。在70年代末和80年代初，这个领域有很多优秀的竞争对手，我想吸取他们的长处并提出新的尝试，因为大环境已经变了：Internet 是高度并发且有高度的安全意识的，因此，偏向于 C 和 C++ 的设计的方案正在发生变化。
```

> > 许多显而易见的好想法，在其他语言中广为人知和喜爱，还没有使其成为广泛使用的系统语言，或者以内存模型非常差（不安全，对并发性敌对）的语言为由而没有进行部署。在70年代末和80年代初，在这个领域有很多优秀的竞争对手，我想根据环境已经改变的理论，复兴他们的一些想法，再给他们一个机会：Internet 是高度并发且有高度的安全意识，因此，用 C 和 C ++ 的设计的方案正在发生变化。

Reviewer 3: ```suggestion
> 许多显而易见的好想法，在其他语言中广为人知和喜爱，还没有使其成为广泛使用的系统语言，或者以内存模型非常差（不安全，对并发性敌对）的语言为由而没有进行部署。在 70 年代末和 80 年代初，在这个领域有很多优秀的竞争对手，我想根据环境已经改变的理论，复兴他们的一些想法，再给他们一个机会：Internet 是高度并发且有高度的安全意识，因此，用 C 和 C ++ 的设计的方案正在发生变化。
```

> > 许多显而易见的好想法，在其他语言中广为人知和喜爱，还没有使其成为广泛使用的系统语言，或者以内存模型非常差（不安全，对并发性敌对）的语言为由而没有进行部署。在70年代末和80年代初，在这个领域有很多优秀的竞争对手，我想根据环境已经改变的理论，复兴他们的一些想法，再给他们一个机会：Internet 是高度并发且有高度的安全意识，因此，用 C 和 C ++ 的设计的方案正在发生变化。

Reviewer: ```suggestion
> 许多显而易见的好想法，在其他语言中被广泛接受和喜爱，却没有加入到被广泛使用的系统语言中，或者只在内存模型非常差（不安全，对并发性敌对）的系统语言中加入。在 70 年代末和 80 年代初，在这个领域有很多优秀的竞争对手，我想根据环境已经改变的理论，复兴他们的一些想法，再给他们一个机会：网络是高并发且对安全很敏感的，因此，一向偏重 C 和 C++ 的设计权衡正在发生变化。
```

> 在 2009 年Mozilla 接受了 Hoare 的工作，并且在 2010 年正式宣布公司成立。在过去的十年中，Rust 蓬勃发展，广受欢迎，并逐渐渗透到为 AWS，Microsoft 和 Google 等公司提供支持的基础架构中。但是它还没有为 Mozilla 带来光明的未来。实际上，[Mozilla 在 2020 年解雇了很大一部分员工](https://www.zdnet.com/article/programming-language-rust-mozilla-job-cuts-have-hit-us-badly-but-heres-how-well-survive/)，其中包括 Rust 的主要贡献者。这些 Rust 贡献者很容易在其他地方找到工作，因为 Rust 对于几乎所有依赖系统工程工作的公司都非常重要。

Reviewer 2: ```suggestion
在 2009 年 Mozilla 接受了 Hoare 的工作，并且在 2010 年正式公开了这一项目。在过去的十年中，Rust 蓬勃发展，广受欢迎，并逐渐渗透到为 AWS，Microsoft 和 Google 等公司提供支持的基础架构中。但是它还没有为 Mozilla 带来光明的未来。实际上，[Mozilla 在 2020 年解雇了很大一部分员工](https://www.zdnet.com/article/programming-language-rust-mozilla-job-cuts-have-hit-us-badly-but-heres-how-well-survive/)，其中包括 Rust 的主要贡献者。这些 Rust 贡献者很容易在其他地方找到工作，因为 Rust 对于几乎所有依赖系统工程工作的公司都非常重要。
```

> 在 2009 年Mozilla 接受了 Hoare 的工作，并且在 2010 年正式宣布公司成立。在过去的十年中，Rust 蓬勃发展，广受欢迎，并逐渐渗透到为 AWS，Microsoft 和 Google 等公司提供支持的基础架构中。但是它还没有为 Mozilla 带来光明的未来。实际上，[Mozilla 在 2020 年解雇了很大一部分员工](https://www.zdnet.com/article/programming-language-rust-mozilla-job-cuts-have-hit-us-badly-but-heres-how-well-survive/)，其中包括 Rust 的主要贡献者。这些 Rust 贡献者很容易在其他地方找到工作，因为 Rust 对于几乎所有依赖系统工程工作的公司都非常重要。

Reviewer 3: ```suggestion
在 2009 年 Mozilla 接受了 Hoare 的工作，并且在 2010 年正式宣布公司成立。在过去的十年中，Rust 蓬勃发展，广受欢迎，并逐渐渗透到为 AWS，Microsoft 和 Google 等公司提供支持的基础架构中。但是它还没有为 Mozilla 带来光明的未来。实际上，[Mozilla 在 2020 年解雇了很大一部分员工](https://www.zdnet.com/article/programming-language-rust-mozilla-job-cuts-have-hit-us-badly-but-heres-how-well-survive/)，其中就包括了不少 Rust 的主要贡献者。这些 Rust 贡献者很容易在其他地方找到工作，因为 Rust 对于几乎所有依赖系统工程工作的公司都非常重要。
```

> 在 2009 年Mozilla 接受了 Hoare 的工作，并且在 2010 年正式宣布公司成立。在过去的十年中，Rust 蓬勃发展，广受欢迎，并逐渐渗透到为 AWS，Microsoft 和 Google 等公司提供支持的基础架构中。但是它还没有为 Mozilla 带来光明的未来。实际上，[Mozilla 在 2020 年解雇了很大一部分员工](https://www.zdnet.com/article/programming-language-rust-mozilla-job-cuts-have-hit-us-badly-but-heres-how-well-survive/)，其中包括 Rust 的主要贡献者。这些 Rust 贡献者很容易在其他地方找到工作，因为 Rust 对于几乎所有依赖系统工程工作的公司都非常重要。

Reviewer: ```suggestion
Mozilla 在 2009 年接受了 Hoare 的工作，并且在 2010 年正式公开了这一项目。在过去的十年中，Rust 蓬勃发展，广受欢迎，并逐渐渗透到为 AWS、Microsoft 和 Google 等公司提供支持的基础架构中。但是它还没有为 Mozilla 带来光明的未来。实际上，[Mozilla 在 2020 年解雇了很大一部分员工](https://www.zdnet.com/article/programming-language-rust-mozilla-job-cuts-have-hit-us-badly-but-heres-how-well-survive/)，其中包括 Rust 的主要贡献者。这些 Rust 贡献者很容易在其他地方找到工作，因为 Rust 对于几乎所有依赖系统工程工作的公司都非常重要。
```

> 让我们把目光转向 Mozilla。 尽管多年来 Mozilla 在技术上取得了令人难以置信的成就，但很难猜测 Mozilla 将会发生什么。Mozilla 令人印象深刻的作品的影响可能很多年都无法完全实现。我们每天或直接或间接依赖的大量云服务越来越多地使用 Rust 构建。

Reviewer 2: ```suggestion
让我们把目光转向 Mozilla 的问题。 尽管多年来 Mozilla 在技术领域做出了让人难以置信的贡献，但  Mozilla 未来仍然扑朔迷离。Mozilla 最令人印象深刻的作品的影响可能在接下来的很多年都无法被完全认识到。我们每天或直接或间接依赖的大量云服务越来越多地使用 Rust 构建。
```

> 让我们把目光转向 Mozilla。 尽管多年来 Mozilla 在技术上取得了令人难以置信的成就，但很难猜测 Mozilla 将会发生什么。Mozilla 令人印象深刻的作品的影响可能很多年都无法完全实现。我们每天或直接或间接依赖的大量云服务越来越多地使用 Rust 构建。

Reviewer: ```suggestion
这又让我们把目光重新聚焦到了 Mozilla 的问题。尽管多年来 Mozilla 在技术领域做出了让人难以置信的贡献，但它的未来仍然扑朔迷离。对于 Mozilla 最令人印象深刻的贡献所带来的影响，人们可能在接下来的很多年都无法完全意识到。我们每天或直接或间接依赖的大量云服务越来越多地使用 Rust 构建。
```

> 在谈到 [Rust 的日益普及](https://redmonk.com/sogrady/2021/03/01/language-rankings-1-21/)时，RedMonk 分析师 James Governor 强调了 Rust 填补各种市场的能力是其成功的关键：『我首先在物联网方面遇到了它­­ -- ­即用于设备编程的 Rust。 显然，它正在作为一种系统编程语言而增长，Rust 和 WASM/WASI 周围的生态系统以及来自 Fastly 的无服务器计算看起来非常有趣。』

Reviewer: ```suggestion
在谈到 [Rust 的日益普及](https://redmonk.com/sogrady/2021/03/01/language-rankings-1-21/)时，RedMonk 分析师 James Governor 强调了 Rust 填补各种市场的能力是其成功的关键：『我首先在物联网领域遇到了它­­ —— 即用于设备编程的 Rust。显然，它作为一种系统编程语言正在成长，Rust 和 WASM/WASI 周围的生态系统以及来自 Fastly 的无服务器计算看起来非常有趣。』
```

> 正如 Mozilla 所[建议](https://research.mozilla.org/rust/)的那样，这种使[开发人员](https://www.techrepublic.com/article/how-to-become-a-developer-a-cheat-sheet/)能够构建『雄心勃勃，快速且正确的』代码的能力几乎可以肯定，它将在系统开发中变得越来越普遍。Mozilla 可能不会直接从这项创新中受益，但是通过 Rust 的发展和贡献，Mozilla 为我们提供了比 Firefox 更大甚至更具战略意义的东西。

Reviewer: ```suggestion
正如 Mozilla 所[表明](https://research.mozilla.org/rust/)的那样，这种使[开发人员](https://www.techrepublic.com/article/how-to-become-a-developer-a-cheat-sheet/)能够构建“健壮、快速且正确的”代码的能力，将让 Rust 在系统开发中变得越来越普遍。Mozilla 可能不会直接从这项创新中受益，但是通过 Rust 的发展和贡献，Mozilla 为我们提供了比 Firefox 更大甚至更具战略意义的东西。
```
